### PR TITLE
Wasm shims detection

### PR DESCRIPTION
**Pull request by flavio**
#### Proposed Changes

As developers start to experiment building WebAssembly applications, they desire to deploy these workloads on top of Kubernetes, using the same paradigm of "traditional" applications.
The [`containerd/runwasi`](https://github.com/containerd/runwasi) project provides a way to integrate these workloads into Kubernetes.

The WebAssembly workloads require dedicated containerd runtimes to be installed and configured on the node. This PR provides a way to automatically discover the already installed runtimes and properly configure containerd to consume them.

The PR doesn't focus on shipping the pre-built runtimes as part of the k3s installation. I think this should be done by another dedicated PR.

#### Types of Changes

This PR introduces a new feature. It does that by leveraging the same mechanism used to discover the nvidia runtimes. As you will see, the existing nvidia discovery code has been refactored. Now we have a generic discovery mechanism that can be reused to discover any kind of containerd runtime.

This can be useful not only to add the WebAssembly runtimes, but to discover and configure other runtimes such as gVisor, Kata, ...

#### Verification

On a node where k3s agent is about to be installed create the following files:

- `/usr/bin/containerd-shim-lunatic-v1`
- `/usr/sbin/containerd-shim-spin-v1`
- `/opt/kwasm/bin/containerd-shim-lunatic-v1`
- `/usr/bin/nvidia-container-runtime`
- `/usr/bin/nvidia-container-runtime-experimental`

All these files can be created with a `touch` command, they must have the executable permission set.

Install the k3s agent, check the containerd `config.toml`. Each runtime should have a dedicated section inside of the file.

#### Testing

The new code is covered by dedicated unit tests.

#### Linked Issues

This PR partially fixes https://github.com/k3s-io/k3s/issues/8700 . This PR doesn't handle the registration of the `RuntimeClass` resource for each of the runtimes discovered.

#### User-Facing Change

```release-note
automatic discovery of WebAssembly runtimes
```

#### Further Comments

This PR is somehow related with https://github.com/k3s-io/k3s/pull/6872 . I've discussed with @brandon an alternative approach to his orignal PR. The idea is to focus on the discovery phase of the WebAssembly runtimes. We can expect the user has somehow installed them on the nodes. We will create another PR later on that changes the build scripts of k3s to ensure the most relevant runtimes are pre-installed with k3s.

### CODE REVIEW

**Comment by flavio** [comment]
CC @vitorsavian

File: pkg/agent/containerd/config_linux.go [hunk 1]
```diff
@@ -60,6 +67,10 @@ func setupContainerdConfig(ctx context.Context, cfg *config.Node) error {
 		cfg.AgentConfig.Systemd = !isRunningInUserNS && controllers["cpuset"] && os.Getenv("INVOCATION_ID") != ""
 	}
 
+	extraRuntimes := make(map[string]templates.ContainerdRuntimeConfig)
+	mergeMaps(extraRuntimes, findNvidiaContainerRuntimes(os.DirFS(string(os.PathSeparator))))
+	mergeMaps(extraRuntimes, findWasiRuntimes(os.DirFS(string(os.PathSeparator))))
```
Comment (brandond) [inline pkg/agent/containerd/config_linux.go:72]: Just make the `findXRuntimes` functions take the map, and add the entries to it directly. No need to add an extra function to merge maps. 
```suggestion
	runtimeConfigs := make(map[string]templates.ContainerdRuntimeConfig)
	findNvidiaContainerRuntimes(runtimeConfigs, os.DirFS(string(os.PathSeparator)))
	findWasiRuntimes(runtimeConfigs, os.DirFS(string(os.PathSeparator)))
```

File: pkg/agent/containerd/config_linux.go [hunk 2]
```diff
@@ -38,6 +38,13 @@ func getContainerdArgs(cfg *config.Node) []string {
 	return args
 }
 
+// TODO: drop once we switch to go1.21 - replace with https://pkg.go.dev/maps#Copy
+func mergeMaps(dst, src map[string]templates.ContainerdRuntimeConfig) {
+	for key, value := range src {
+		dst[key] = value
+	}
+}
+
```
Comment (brandond) [inline pkg/agent/containerd/config_linux.go:47]: Drop this, just have the functions add the entries to the existing map
```suggestion
// TODO: drop once we switch to go1.21 - replace with https://pkg.go.dev/maps#Copy
func mergeMaps(dst, src map[string]templates.ContainerdRuntimeConfig) {
	for key, value := range src {
		dst[key] = value
	}
}

```

File: pkg/agent/containerd/config_linux.go (hunk 1 above)
Comment (brandond) [reply pkg/agent/containerd/config_linux.go:72]: Could probably also do this to make the function signature smaller

```golang
type runtimeConfigs map[string]templates.ContainerdRuntimeConfig
```

**Review by brandond — CHANGES_REQUESTED** [review]

**Review by vitorsavian — APPROVED** [review]

File: pkg/agent/containerd/config_linux.go (hunk 2 above)
Comment (flavio) [reply pkg/agent/containerd/config_linux.go]: done

File: pkg/agent/containerd/config_linux.go (hunk 1 above)
Comment (flavio) [reply pkg/agent/containerd/config_linux.go]: I've created the alias you proposed and changed the `findXRuntimes` to take a map as one of the params

**Comment by flavio** [comment]
@brandond thanks for the feedback, I've applied the changes you proposed and squashed them into the last commit

File: pkg/agent/containerd/nvidia.go [hunk 3]
```diff
@@ -4,20 +4,17 @@
...
 
 // findNvidiaContainerRuntimes returns a list of nvidia container runtimes that
 // are available on the system. It checks install locations used by the nvidia
 // gpu operator and by system package managers. The gpu operator installation
 // takes precedence over the system package manager installation.
 // The given fs.FS should represent the filesystem root directory to search in.
-func findNvidiaContainerRuntimes(root fs.FS) map[string]templates.ContainerdRuntimeConfig {
+func findNvidiaContainerRuntimes(root fs.FS,
+	foundRuntimes runtimeConfigs,
+) {
```
Comment (dereknola) [inline pkg/agent/containerd/nvidia.go:17]:
```suggestion
func findNvidiaContainerRuntimes(root fs.FS, foundRuntimes runtimeConfigs) {
```

File: pkg/agent/containerd/runtimes.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// are available on the system. It checks install locations provided via
+// the potentialRuntimes variable.
+// The binaries are searched at the locations specivied by locationsToCheck.
+// Note: check the given locations in order.
+// The given fs.FS should represent the filesystem root directory to search in.
+func findContainerRuntimes(root fs.FS,
+	potentialRuntimes runtimeConfigs,
+	locationsToCheck []string,
+	foundRuntimes runtimeConfigs,
+) {
```
Comment (dereknola) [inline pkg/agent/containerd/runtimes.go:29]:
```suggestion
func findContainerRuntimes(root fs.FS, 	potentialRuntimes runtimeConfigs, locationsToCheck []string, foundRuntimes runtimeConfigs) {
```

File: pkg/agent/containerd/runtimes_test.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+package containerd
+
+import (
+	"io/fs"
+	"reflect"
+	"testing"
+	"testing/fstest"
+)
+
+func Test_UnitFindRuntimes(t *testing.T) {
```
Comment (dereknola) [inline pkg/agent/containerd/runtimes_test.go:13]:
```suggestion
func Test_UnitFindContainerRuntimes(t *testing.T) {
```

Comment (dereknola) [reply pkg/agent/containerd/runtimes_test.go:13]: Naming for unit tests should follow standard format https://github.com/k3s-io/k3s/blob/master/tests/TESTING.md#format

File: pkg/agent/containerd/runwasi_test.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+package containerd
+
+import (
+	"io/fs"
+	"reflect"
+	"testing"
+	"testing/fstest"
+)
+
+func Test_UnitFindWasiContainerRuntimes(t *testing.T) {
```
Comment (dereknola) [inline pkg/agent/containerd/runwasi_test.go:13]:
```suggestion
func Test_UnitFindWasiRuntimes(t *testing.T) {
```

File: pkg/agent/containerd/runwasi_test.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		args args
+		want runtimeConfigs
+	}{
+		{
+			name: "No runtimes",
+			args: args{
+				root: fstest.MapFS{},
+			},
+			want: runtimeConfigs{},
+		},
```
Comment (dereknola) [inline pkg/agent/containerd/runwasi_test.go:29]:
```suggestion
	tests := []struct {
		name string
		arg fs.FS
		want runtimeConfigs
	}{
		{
			name: "No runtimes",
			arg: fstest.MapFS{},
			want: runtimeConfigs{},
		},
```

Comment (dereknola) [reply pkg/agent/containerd/runwasi_test.go:29]: When only a single argument exists, no need to embed it into a struct.

File: pkg/agent/containerd/runwasi_test.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+			foundRuntimes := runtimeConfigs{}
+			findWasiRuntimes(tt.args.root, foundRuntimes)
+			if !reflect.DeepEqual(foundRuntimes, tt.want) {
+				t.Errorf("findWasiRuntimes() = %+v\nWant = %+v", foundRuntimes, tt.want)
+			}
+		})
+	}
+}
+
+func Test_UnitFindWasiContainerRuntimesPreserveFoundRuntimes(t *testing.T) {
```
Comment (dereknola) [inline pkg/agent/containerd/runwasi_test.go:169]: This is still testing the `findWasiRuntimes` function and thus should be merged with the above unit test. Implementing a `name` field similar to Test_UnitFindRuntimes would be most preferred.

**Review by dereknola — CHANGES_REQUESTED** [review]

File: pkg/agent/containerd/runwasi_test.go (hunk 8 above)
Comment (flavio) [reply pkg/agent/containerd/runwasi_test.go]: I've done that via e4d26eb086f869e3486ec33ae74daecd9faf1def

File: pkg/agent/containerd/runwasi_test.go (hunk 7 above)
Comment (flavio) [reply pkg/agent/containerd/runwasi_test.go]: I've now added another arg, see e4d26eb086f869e3486ec33ae74daecd9faf1def

**Comment by flavio** [comment]
@dereknola I've applied your suggestions and performed the changes you requested. Can you take another look please?

Before merging the PR I'll clean the commit history, I'm talking about the last 2 commits which I left around to simplify the review process

**Comment by dereknola** [comment]
Drone failed because of linter
```
pkg/agent/containerd/runtimes.go:25: File is not `gofmt`-ed with `-s` (gofmt)
```

**Comment by flavio** [comment]
I've fixed the gofmt issue and cleaned up the commit history

**Review by brandond — APPROVED** [review]

**Comment by flavio** [comment]
Nice! I'll put aside some time to update the docs

**Pull request merged**